### PR TITLE
subsys: fb: add support for generating CFB font headers from TrueType/OpenType/image files

### PR DESCRIPTION
**Pull request by henrikbrixandersen**
Add script and cmake functions for automatically generating Character Frame Buffer (CFB) font header files from image files, TrueType, or OpenType font files.

Also add a sample application for demonstrating automatic generation of font header files at build time.

### CODE REVIEW

**Review by dbkinder — COMMENTED** [review]
This patch seems very specific to the reel board, but impacts build requirements for the whole project  (e.g., the ``scripts/requirements.txt`` addition of Pillow).  Is there a GitHub repo for the reel board demo?  That seems a better place for this material rather than in the Zephyr documentation?

**Comment by henrikbrixandersen** [comment]
The patch is not specific to the reel_board. It should work for any board with a display supporting the framebuffer subsystem (but I have only verified it on a reel_board). 

I agree that the added dependency is unfortunate. The “reel_board demo” is in the tree but does not add any new build dependencies.

**Comment by henrikbrixandersen** [comment]
Shippable fails due to the added dependency on Pillow.

**Review by pfalcon — DISMISSED** [review]
@henrikbrixandersen, thanks for the patch! But for a moment I thought I missed to switch tabs and am looking at a different project, because connection between Zephyr and this patch isn't much obvious from the description.

@henrikbrixandersen, if this patch is for, and benefits, **subsys/fb**, then it should be stated and described clearly. For example, even if it happens that you patch only cmake files, but the logical beneficiary of these change is subsys/fb, then I would argue that the patch title should start with `subsys: fb: `, not with `cmake:`.

Thanks.

**Comment by henrikbrixandersen** [comment]
@dbkinder I have reorganised the sample application a bit to make it less specific to the reel_board. I still do not have access to any other boards with onboard display, so the sample has only been validated with that board.

**Comment by henrikbrixandersen** [comment]
@pfalcon Thanks. I have reworded the commit messages as suggested.

**Comment by pfalcon** [comment]
@henrikbrixandersen: Thanks! Not familiar enough with that area to give +1, but dismissing my old review.

**Comment by carlescufi** [comment]
@SebastianBoe could you take a look at the CMake side of things?
@jfischer-phytec-iot @vanwinkeljan could you review the display/framebuffer code?

**Review by carlescufi — COMMENTED** [review]
I don't have enough knowledge in the area to review this, but in general it looks sane.

**Review by vanwinkeljan — APPROVED** [review]
CFB related code lgtm

File: samples/display/cfb_custom_font/README.rst [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+Custom Fonts
+############
+
+Overview
+********
+A simple example showing how to generate Character Framebuffer (CFB)
+font headers automatically at build time.
+
+This example generates a font with font elements for 6 sided dice from
+a PNG image.
```
Comment (dbkinder) [inline samples/display/cfb_custom_font/README.rst]: Add...
```
a PNG image, and then uses the generated header (``cfb_font_dice.h``)
to show the font elements on the display of a supported board.
```

Comment (dbkinder) [reply samples/display/cfb_custom_font/README.rst]: Build and run instructions would be nice.

**Review by dbkinder — CHANGES_REQUESTED** [review]
I yield to the development team wisdom on adding a new build dependency (Pillow) to the project.

**Comment by pfalcon** [comment]
> @dbkinder We currently only really have one requirements.txt, and I don't think we should consider splitting this up at this point just because an additional Python module is required.

Actually, I raised a question quite ago that `requirements.txt` should be split. Mere humans among us won't build documentation, so should not be burdened with the need to install such dependencies. And vice-versa, those who would be built, almost certainly will do that in virtualenv. So, now I need to go and edit (filter) requirements.txt manually to use it virtualenv (because tools for normal build come from OS packages at all, and I don't want to have sudden version jumps because something unneeded for docs got installed in venv).

So, single `requirements.txt` == more burden for everyone out there (and impressions that Zephyr does sloppy dependencies management).

File: scripts/gen_cfb_font_header.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+ *
+ */
+
+#ifndef %s
+#define %s
+
+#include <zephyr.h>
+#include <display/cfb.h>
+
+const u8_t cfb_font_%s_%d%d[%d][%d] = {\n""" % (" ".join(sys.argv),
```
[the remark this answers is not in the document]
Comment (henrikbrixandersen) [reply scripts/gen_cfb_font_header.py]: @carlescufi Thanks for the suggestion. However, in this particular situation, I just want a space separated string containing all the arguments passed to the program to let the user know how a given header file was generated in the first place.

I have, however, converted all occurances of % to .format() and in turn made the string substitutions more readable.

File: samples/display/cfb_custom_font/README.rst (hunk 1 above)
Comment (henrikbrixandersen) [reply samples/display/cfb_custom_font/README.rst]: @dbkinder Thank you. I have added the suggested improvement to the general description of the sample application and added build instructions.

**Review by dbkinder — APPROVED** [review]
+1 for docs.  I'll leave the requirements.txt dependency discussion to you developers :)
Note that the CI testing for this PR is showing an error because it doesn't know about the new requirements.txt dependency this PR creates.

**Comment by henrikbrixandersen** [comment]
@carlescufi ping

**Comment by carlescufi** [comment]
@henrikbrixandersen can you please rebase?

**Comment by henrikbrixandersen** [comment]
@carlescufi Rebased.

**Comment by carlescufi** [comment]
@henrikbrixandersen can you fix CI?

**Review by jfischer-no — CHANGES_REQUESTED** [review]
Looks familiar, but something is missing, I need a little time to take a closer look.

**Comment by henrikbrixandersen** [comment]
@jfischer-phytec-iot Did you notice the added python module dependency on Pillow?

**Review by nashif — CHANGES_REQUESTED** [review]

File: cmake/extensions.cmake [hunk 3]
```diff
@@ -516,6 +517,72 @@ function(generate_inc_file_for_target
   generate_inc_file_for_gen_target(${target} ${source_file} ${generated_file} ${generated_target_name} ${ARGN})
 endfunction()
 
+# 1.4 generate_cfb_font_*
```
[the remark this answers is not in the document]
Comment (henrikbrixandersen) [reply cmake/extensions.cmake]: @nashif Thank you for the review. I have put the gen_cfb_* functions into cmake/cfb.cmake instead. Please let me know, if I should use another filename for it.

File: scripts/gen_cfb_font_header.py [hunk 4]
```diff
@@ -0,0 +1,157 @@
+#!/usr/bin/env python3
```
[the remark this answers is not in the document]
Comment (henrikbrixandersen) [reply scripts/gen_cfb_font_header.py]: @nashif Thanks. I have addressed most of them now:

$ pylint scripts/gen_cfb_font_header.py
************* Module gen_cfb_font_header
scripts/gen_cfb_font_header.py:1:0: C0111: Missing module docstring (missing-docstring)
scripts/gen_cfb_font_header.py:110:4: C0103: Constant name "args" doesn't conform to UPPER_CASE naming style (invalid-name)
scripts/gen_cfb_font_header.py:110:4: W0601: Global variable 'args' undefined at the module level (global-variable-undefined)

------------------------------------------------------------------
Your code has been rated at 9.57/10 (previous run: 9.57/10, +0.00)

OK?

[the remark this answers is not in the document]
Comment (nashif) [reply scripts/gen_cfb_font_header.py]: 👍 Awesome :)

**Review by nashif — APPROVED** [review]

**Comment by henrikbrixandersen** [comment]
@jfischer-phytec-iot ping?

File: scripts/gen_cfb_font_header.py [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+
+def generate_element(image, charcode):
+    """Generate CFB font element for a given character code from an image"""
+    blackwhite = image.convert("1", dither=Image.NONE)
+    pixels = blackwhite.load()
+
+    if args.dump:
+        blackwhite.save("{}_{}.png".format(args.name, charcode))
+
+    args.output.write("""\t/* {:d} */\n\t{{\n""".format(charcode))
```
Comment (jfischer-phytec-iot) [inline scripts/gen_cfb_font_header.py]: I suggest to print the character here.
```diff
-    args.output.write("""\t/* {:d} */\n\t{{\n""".format(charcode))
+    args.output.write("""\t/* {:c} */\n\t{{\n""".format(charcode))
```

File: scripts/gen_cfb_font_header.py [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+        extract_image_glyphs()
+    elif args.input.name.lower().endswith((".otf", ".otc", ".ttf", ".ttc")):
+        extract_font_glyphs()
+    else:
+        extract_image_glyphs()
+
+    args.output.write("""
+}};
+
+FONT_ENTRY_DEFINE({name},
```
Comment (jfischer-phytec-iot) [inline scripts/gen_cfb_font_header.py]: `name` should be uniquely, maybe {name}_{width}{height}?

**Review by jfischer-no — COMMENTED** [review]
Thank you and sorry for the delay, everything looks really good, just two minor comments

**Comment by jfischer-phytec-iot** [comment]
@henrikbrixandersen [Here](https://gist.github.com/jfischer-phytec-iot/5b133a91f603dd392ca3223738159fa7) is the script I used to to generate `subsys/fb/cfb_fonts.c` (I did not have time to improve it)

File: scripts/gen_cfb_font_header.py (hunk 6 above)
Comment (henrikbrixandersen) [reply scripts/gen_cfb_font_header.py]: @jfischer-phytec-iot Good catch. Fixed. I must have accidentally changed that during one of the rewrites.

File: scripts/gen_cfb_font_header.py (hunk 5 above)
Comment (henrikbrixandersen) [reply scripts/gen_cfb_font_header.py]: @jfischer-phytec-iot I did that in an initial revision, but it only works for ASCII printable character codes. The font generator does not restrict you to use printable characters.

Comment (jfischer-phytec-iot) [reply scripts/gen_cfb_font_header.py]: Agree, but it is very handy for the fonts, can we have both formats {:d} and {:c}? :smile:

**Review by jfischer-no — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**